helix-tui/src/buffer.rs

```diff
@@ -289,28 +292,60 @@ impl Buffer {
         let width = if ellipsis { width - 1 } else { width };
         let graphemes = UnicodeSegmentation::graphemes(string.as_ref(), true);
         let max_offset = min(self.area.right() as usize, width.saturating_add(x as usize));
-        for s in graphemes {
-            let width = s.width();
-            if width == 0 {
-                continue;
+        if !truncate_start {
+            for s in graphemes {
+                let width = s.width();
+                if width == 0 {
+                    continue;
+                }
+                // `x_offset + width > max_offset` could be integer overflow on 32-bit machines if we
+                // change dimenstions to usize or u32 and someone resizes the terminal to 1x2^32.
+                if width > max_offset.saturating_sub(x_offset) {
+                    break;
+                }
+
+                self.content[index].set_symbol(s);
+                self.content[index].set_style(style);
+                // Reset following cells if multi-width (they would be hidden by the grapheme),
+                for i in index + 1..index + width {
+                    self.content[i].reset();
+                }
+                index += width;
+                x_offset += width;
             }
-            // `x_offset + width > max_offset` could be integer overflow on 32-bit machines if we
-            // change dimenstions to usize or u32 and someone resizes the terminal to 1x2^32.
-            if width > max_offset.saturating_sub(x_offset) {
-                break;
+            if ellipsis && x_offset - (x as usize) < string.as_ref().width() {
+                self.content[index].set_symbol("…");
             }
-
-            self.content[index].set_symbol(s);
-            self.content[index].set_style(style);
-            // Reset following cells if multi-width (they would be hidden by the grapheme),
-            for i in index + 1..index + width {
-                self.content[i].reset();
+        } else {
+            let mut truncated = false;
+            let mut end = Vec::new();
```

Comment:
I think we can do better than collecting iterating over this twice and having a separate allocation.

I think probably can just get the index and display the rest. Didn't look closely.

Reply:
Since individual graphemes can have different widths, I'm not sure how to do this without iterating over it in reverse and checking the total width.

> single character for each directory in path

Ah, that's a good idea.

Reply:
You should be able to just calculate the total width using `unicode_width::UnicodeWidthStr`, no need to segment it into graphemes: https://github.com/helix-editor/helix/blob/e505bf2b4809681840d1684a23ad21255fca290d/helix-core/src/graphemes.rs#L30

We re-export the library as `helix_core::unicode::width`

Reply:
Ah yeah, it's the same method you're already using (`.width()`). Should be enough to call  `string.width()`, then if it's larger than `width` we iterate graphemes from reverse, rendering them into `content` (without using the `end` vector)

Reply:
I pushed a rewrite with ^ but it's untested. Probably needs more work.

Reply:
Yeah… this probably isn't what's supposed to happen 😄 
<img width="364" alt="image" src="https://user-images.githubusercontent.com/79615262/140236895-ab9864b1-7973-4484-83ee-46e3e38358fa.png">

Reply:
And it panics when the paths are long enough to actually need to be truncated.

Reply:
Ah, addressing [this](https://github.com/helix-editor/helix/pull/951#discussion_r741586853) fixes the panic.

Reply:
Ah, I guess that codepath always gets called even if the string is shorter.

Reply:
We'd preferably use the original code under `!truncate_start` if the width is shorter

Reply:
I think I've figured out what to do.